mesonbuild/modules/codegen.py

```diff
@@ -0,0 +1,230 @@
+        if bin.name == 'win_flex':
+            args.append('--wincompat')
+        args.extend(['-o', '@OUTPUT0@', '@INPUT@'])
+        self._generators['lex'] = Generator(bin, T.cast('ImmutableListProtocol[str]', args))
+
+    @typed_pos_args('codegen.lex', (str, File, GeneratedList, CustomTarget, CustomTargetIndex))
+    @typed_kwargs(
+        'codegen.lex',
+        KwargInfo('version', ContainerTypeInfo(list, str), default=[], listify=True),
+        KwargInfo('args', ContainerTypeInfo(list, str), default=[], listify=True),
+        KwargInfo('source', (str, NoneType)),
+        KwargInfo('header', (str, NoneType)),
+        KwargInfo('table', (str, NoneType)),
+        KwargInfo('plainname', bool, default=False),
+    )
+    def lex_method(self, state: ModuleState, args: T.Tuple[T.Union[str, File, GeneratedList, CustomTarget, CustomTargetIndex]], kwargs: LexKwargs) -> ModuleReturnValue:
+        self.__find_lex(state, kwargs['version'])
+
+        input = state._interpreter.source_strings_to_files([args[0]])[0]
+        if isinstance(input, File):
+            is_cpp = input.endswith(".ll")
+            name = os.path.splitext(input.fname)[0]
+        else:
+            is_cpp = input.get_outputs()[0].endswith('.ll')
+            name = os.path.splitext(input.get_outputs()[0])[0]
+        name = os.path.basename(name)
+
+        if is_cpp and 'flex' in self._generators['lex'].program.name:
+            # Flex uses FlexLexer.h for C++ code
+            try:
+                found = state.environment.coredata.compilers.host['cpp'].has_header('FlexLexer.h', '', state.environment)
+                if not found[0]:
+                    raise MesonException('Could not find FlexLexer.h, which is required for Flex with C++')
+            except KeyError:
+                raise MesonException("Could not find a C++ compiler to search for FlexLexer.h")
+
+        base = '@PLAINNAME@' if kwargs['plainname'] else '@BASENAME@'
+        if kwargs['source'] is None:
+            outputs = [f'{base}.{"cpp" if is_cpp else "c"}']
+        else:
+            outputs = [kwargs['source']]
+
+        command = self._generators['lex'].command()
+        if kwargs['header'] is not None:
+            outputs.append(kwargs['header'])
+            command.append(f'--header-file=@OUTPUT{len(outputs)}@')
+        if kwargs['table'] is not None:
+            outputs.append(kwargs['table'])
+            command.append(f'--tables-file=@OUTPUT{len(outputs)}@')
+        command.extend(kwargs['args'])
+
+        target = CustomTarget(
+            f'codegen-lex-{name}',
+            state.subdir,
+            state.subproject,
+            state.environment,
+            command,
+            [input],
+            outputs,
+            backend=state.backend,
+            description='Generating lexer {} with lex',
+        )
+
+        return ModuleReturnValue(target, [target])
+
+    def __find_yacc(self, state: ModuleState, version: T.List[str]) -> None:
```

Comment:
Do we have a way to specify "no, I really need flex+bison extensions, the POSIX variants aren't good here"?

Reply:
From inside the meson.build, no. But that actually probably makes sense. I had though that would be a machine file issue